File: bzoj/1231/1231.py
import Queue

input = open('1231.in', 'r')
line = input.readline().split()

class Node :
	def __init__(self, a, b):
		self.last = a
		self.stat = b

n = int(line[0])
K = int(line[1])
i = 0
j = 0
num = [i for i in range(20)]
vis = [[0] * 20] * 135550
dp = [[0] * 20] * 135550
q = Queue.Queue()
for i in range(n) :
	line = input.readline().split()
	num[i] = int(line[0])
	q.put(Node(i, 1 << i))
	dp[i][1 << i] = 1
	vis[i][1 << i] = 1
input.close();

while not q.empty() :
	cur = q.get()
	vis[cur.last][cur.stat] = 0

	for i in range(n) :
		if ((1 << i) & cur.stat) > 0 : continue
		if (abs(num[i] - num[cur.last]) <= K) :
			continue

		nex = (cur.stat | (1 << i))
		dp[i][nex] += dp[cur.last][cur.stat]

		if vis[i][nex] == 0 :
			vis[i][nex] = 1
			print i, nex
			q.put(Node(i, nex))
ans = 0
for i in range(n) :
	ans += dp[i][(1 << n) - 1]
print ans

File: bzoj/1231/1231.cpp
#include <bits/stdc++.h>

using namespace std;

struct Node{
	int last, stat;
	long long cnt;
	Node() {}
	Node(int a, int b) : last(a), stat(b) {}
};

int n, K;
int num[20];
long long dp[20][135550];
bool vis[20][135550];

int main() {
	freopen("1231.in", "r", stdin);
	scanf("%d %d", &n, &K);
	queue<Node> q;
	for (int i = 0; i < n; i++) {
		scanf("%d", &num[i]);
		q.push(Node(i, 1 << i));
		dp[i][1 << i] = 1;
		vis[i][1 << i] = 1;
	}
	while (!q.empty()) {
		Node cur = q.front();
		q.pop();
		vis[cur.last][cur.stat] = 0;

		for (int i = 0; i < n; i++) {
			if ((1 << i) & cur.stat) continue;
			if (abs(num[i] - num[cur.last]) <= K) continue;
			int nex = cur.stat | (1 << i);
			dp[i][nex] += dp[cur.last][cur.stat];

			if (!vis[i][nex]) {
				vis[i][nex] = 1;
				cout << i << ' ' << nex << endl;
				q.push(Node(i, nex));
			}
		}
	}
	long long ans = 0;
	for (int i = 0; i < n; i++)
		ans += dp[i][(1 << n) - 1];
	printf("%lld\n", ans);

	return 0;
}
